extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/StartupBuildSteps.java

```diff
@@ -175,44 +178,67 @@ private void registerStartupObserver(ObserverRegistrationPhaseBuildItem observer
         configurator.id(id);
         configurator.priority(priority);
         configurator.notify(mc -> {
+            BytecodeCreator bytecode = mc;
+
             // InjectableBean<Foo> bean = Arc.container().bean("bflmpsvz");
-            ResultHandle containerHandle = mc.invokeStaticMethod(ARC_CONTAINER);
-            ResultHandle beanHandle = mc.invokeInterfaceMethod(ARC_CONTAINER_BEAN, containerHandle,
-                    mc.load(bean.getIdentifier()));
+            ResultHandle containerHandle = bytecode.invokeStaticMethod(ARC_CONTAINER);
+            ResultHandle beanHandle = bytecode.invokeInterfaceMethod(ARC_CONTAINER_BEAN, containerHandle,
+                    bytecode.load(bean.getIdentifier()));
+
+            // if the [synthetic] bean is not active and is not injected in a user bean, ignore `InactiveBeanException`
+            // this means that an inactive bean that is injected into a user bean will end up with an error
+            if (bean.hasCheckActiveMethod()) {
+                boolean isInjectedInUserBean = false;
+                for (InjectionPointInfo ip : observerRegistration.getBeanProcessor().getBeanDeployment().getInjectionPoints()) {
+                    BeanInfo resolvedBean = ip.getResolvedBean();
+                    if (resolvedBean != null && !resolvedBean.isSynthetic() && resolvedBean.equals(bean)) {
```

Comment:
I noticed injecting a deactivated datasource into a user bean didn't fail startup, so I looked at this condition and... I think the condition `resulvedBean.isSynthetic()` is wrong here?

If `resolvedBean.equals(bean)`, then the resolved bean is the bean with a `checkActive` method, so it is necessarily synthetic (only way to add a `checkActive` method right now). So `!resolvedBean.isSynthetic() && resolvedBean.equals(bean)` will always be `false`.

I think we want to check that the bean *being injected with `resolvedBean`* is non-synthetic, right? So the condition should rather be `!ip.getTargetBean().map(BeanInfo::isSynthetic).orElse(false)`. Or something like that: I'm not sure what we want to do when there is no target bean, because what does that even mean? :thinking:

Reply:
Ouch. Good point, I guess I didn't think too much about this. You're right, the `isSynthetic()` method should be called on the _target_ bean of the IP (which is the bean that declares the IP), not on the resolved bean (which is the bean the IP resolves to).

Reply:
Done.